tools/report_build.py

```diff
@@ -125,23 +188,85 @@ class MinimizeComment(GitHubQLQuery):
     """
 
 
+class GitHubV3Call():
+    variables = {}
+
+    def __init__(self, variables={}):
+        self.github = Github(os.environ["GITHUB_TOKEN"])
+        self.variables = variables
+
+    def run(self, variables={}):
+        raise NotImplementedError(
+            "You need to subclass GitHubV3Call and"
+            " provide the run(self, variables={}) method"
+        )
+
+    def _vars(self, variables):
+        return dict_merge(self.variables, variables)
+
+
+class RepoCall(GitHubV3Call):
+    def run(self, variables={}):
+        self._repo = self.github.get_repo(
+            "{owner}/{name}".format(**self._vars(variables)))
+
+
+class AddCommitComment(RepoCall):
+    def run(self, variables={}):
+        super().run(variables)
+        return (self._repo.get_commit(self._vars(variables)["sha"])
+                .create_comment(self._vars(variables)["comment"]["body"]))
+
+
+class UpdateCommitComment(RepoCall):
+    def run(self, variables={}):
+        super().run(variables)
+        return (self._repo.get_comment(
+            self._vars(variables)["comment"]["databaseId"])
+            .edit(self._vars(variables)["comment"]["body"]))
+
+
 def main():
     build_name = sys.argv[1]
     comment_body = ["{} build available: {}".format(
         build_name, " or ".join(sys.argv[2:]))]
 
     owner, name = os.environ["TRAVIS_REPO_SLUG"].split("/")
+    travis_event = os.environ["TRAVIS_EVENT_TYPE"]
 
-    events_d = GetEvents().run({
+    common_d = {
         "owner": owner,
         "name": name,
-        "pull_request": int(os.environ["TRAVIS_PULL_REQUEST"]),
-        "count": EVENT_COUNT
-    })["data"]
+    }
+
+    if travis_event == "pull_request":
+        events_d = GetEvents(common_d).run({
+            "pull_request": int(os.environ["TRAVIS_PULL_REQUEST"]),
+            "count": EVENT_COUNT
+        })["data"]
+        add_comment = AddComment(dict(common_d, **{
+            "comment": {
+                "subjectId": events_d["repository"]["pullRequest"]["id"]
+            }
+        }))
+        update_comment = UpdateComment()
+        events = (events_d["repository"]["pullRequest"]
+                  ["timelineItems"]["edges"])
+
+    elif travis_event == "push":
+        events_d = GetCommitComments(common_d).run({
+            "commit": os.environ["TRAVIS_COMMIT"]
+        })["data"]
+        add_comment = AddCommitComment(dict(common_d, **{
+            "sha": os.environ["TRAVIS_COMMIT"]
+        }))
+        update_comment = UpdateCommitComment(common_d)
+        events = events_d["repository"]["object"]["comments"]["edges"]
```

Comment:
I'm not so sure about this approach any more :)

Case in point: #1135 - the `release/0.9` PR. It got built [and published on the commit](https://github.com/CanonicalLtd/multipass/commit/62cb4cbfd10196f7ecea0687cc08f864102d307f#commitcomment-35631219), but then I [force-pushed it](https://github.com/CanonicalLtd/multipass/pull/1135#event-2737239426) to 62cb4cbfd10196f7ecea0687cc08f864102d307f, so the pull request timeline lost that commit, and the comment with the build info along with it.

While it may be logically correct (the PR no longer has that commit, so its builds are not relevant any more), but it means we lose the history.

I could send the comment to all the PRs associated with the commit, or to both… But then we'd have double the comments in the generic case. If we only comment on the PR, the commit itself won't have it.

There should be a way to submit a single commit for both the commit and the PR ;P (but there isn't)

Reply:
@Saviq, I still wonder if there would be a way to just reuse the PR case here. Couldn't this be done through PRs with a different base?

Reply:
As discussed on IRC, when [building pull requests](https://docs.travis-ci.com/user/pull-requests/#how-pull-requests-are-built) Travis (rightly so) builds a merge of the PR'd branch with the base (basically `+refs/pull/<number>/merge`.

So the first thing we'd need to do is to `git checkout ${TRAVIS_PULL_REQUEST_BRANCH}` or so. The way it is now here, in https://github.com/CanonicalLtd/multipass/pull/1135#partial-pull-merging you can see it actually built both - `push` and `pr` - the merge and HEAD, respectively. Both have merit.

Also that's the only way to build on tags (Travis will build the same commit a second time if it's tagged - in a tag context).

So the only question I'm asking here is where do the builds "belong", and I'm thinking that it depends on context… If it's a plain branch build, it's the commits that get built, so it makes sense to comment on the commit. But if we're building RCs, it's only an RC in the branch context, not on its own. And a PR has a much closer connection with a branch than a commit does (as demonstrated above, where we "lost" a relevant comment from the PR timeline).

I have to check a couple of things in relation to commits and their associated pull requests in the API, but I'm thinking at least in the RC case I will "publish" on the PR rather than the commit.

I'll leave the other branch builds as they are today, so that we get builds for `master` and `bors try`.

Reply:
And now that's implemented.